glusterfs service: add support for TLS communication

bachp opened the pull request:
###### Motivation for this change

This allows to easily setup glusterfs with TLS support. The user has to provide three files: 
- Private Key
- Certificate
- Certificate Authority

There are still some open points I would appreciate some feedback:

1. Enabling TLS in GlusterFS works by placing the certificates in `/etc/ssl` and creating the marker file `/var/lib/glusterd/secure-access`. This is the same setup for Servers and Clients wanting to mount from a secure server. This PR only activates this for the server part. This means one can mount a TLS secured volume on a machine where the server is also running. To improve this the TLS option should probably be moved out of the service config, but I'm not sure how.


###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

/cc @nh2

grahamc commented on nixos/modules/services/network-filesystems/glusterfs.nix:
```diff
@@ -5,6 +5,22 @@ with lib;
+
+  restartTriggers = [
+    config.environment.etc."ssl/glusterfs.pem".source
+    config.environment.etc."ssl/glusterfs.key".source
```
Users should probably not put the private key in to the nix store.

bachp replied:
What is a better approach?

nh2 replied:
> What is a better approach?

Giving it a path (that's not in the nix store).

For example, `nixops` has a `keys` functionality that places secret keys into a special directory that doesn't survive reboots.

nh2 commented on nixos/modules/services/network-filesystems/glusterfs.nix:
```diff
@@ -30,6 +46,30 @@ in
+      };
+
+      caCert = mkOption {
+        default = null;
```
There should be some `assert`s here that ensure that when `enableTLS` is true, `tlsKey` and `tlsKey` and `tlsPem` are not null.

Better yet, make a `tlsSettings` option of type `submodule` that can be either `null` (for TLS being disabled), or a `{}` with the 3 options inside, each non-nullable.

An example of how to do that can be found here:

https://github.com/nh2/nixops-gluster-example/blob/8a859f0a702efd38bef9a202e1b031f1db6a44d3/modules/nh2-glusterfs-server.nix#L128-L160

bachp commented on nixos/modules/services/network-filesystems/glusterfs.nix:
```diff
@@ -5,6 +5,22 @@ with lib;
+  ''
+  else
+  ''
+    rm -f /var/lib/glusterd/secure-access
```
If we don't want to add the keys to the nixstore and don't manipulate the file in `/var/lib/glusterd` then there is actually nothing much this change does. As anyway everything (put keys somewhere, put activation file in `/var`) has to be done outside of nix?

nh2 replied:
I think you misunderstood me: I think this PR is good, and that deleting `secure-access` here is OK. You just need to document that behaviour in the option `description`.

Also, having the pubkey in the store is OK, it's just the private key that should be a file path instead.

bachp replied:
Ok thanks for clarifying :)

I would imagine instead of putting the private key into the config I would just put a path to a file.
What's the mechanism to make sure this gets symlinked to the correct location in etc? Also via `config.environment.etc`?

Is there a similar mechanism that could be used to put the file into `/var`?

bachp committed:
glusterfs service: add support for TLS communication

TLS settings are implemented as submodule.

bachp commented:
I reworked the PR to make use of submodule as @nh2 suggested.

joachifm commented:
Is this ready?

bachp commented:
@joachifm from my point it is. I would appreciate some feedback from @nh2 tough.

nh2 commented on nixos/modules/services/network-filesystems/glusterfs.nix:
```diff
@@ -30,6 +46,39 @@ in
+            caCert = mkOption {
+              default = null;
+              type = types.path;
+              description = "Path certificate authority used to signe the cluster certificates.";
```
2 small typos: Path `to` and `signe` -> `sign`.

nh2 commented on nixos/modules/services/network-filesystems/glusterfs.nix:
```diff
@@ -30,6 +46,39 @@ in
+
+      tlsSettings = mkOption {
+        description = ''
+          Make the server communicat via TLS.
```
typo `communicate`

nh2 replied:
I recommend adding a link to the docs: https://gluster.readthedocs.io/en/latest/Administrator%20Guide/SSL/

nh2 commented on nixos/modules/services/network-filesystems/glusterfs.nix:
```diff
@@ -30,6 +46,39 @@ in
+        type = types.nullOr (types.submodule {
+          options = {
+            tlsKey = mkOption {
+              default = null;
```
I think you need to remove this `default = null` and the other two below.

Either we want that the entire `tlsSettings` dict is `null`, or we want that it is not null, and in that case we want that the user provides all 3 of `tlsKey`, `tlsPem` and `caCert`, so those should not be nullable.

nh2 commented on nixos/modules/services/network-filesystems/glusterfs.nix:
```diff
@@ -40,7 +89,14 @@ in
+    environment.etc = mkIf (cfg.tlsSettings != null) {
+      "ssl/glusterfs.pem".source = cfg.tlsSettings.tlsPem;
+      "ssl/glusterfs.key".source = cfg.tlsSettings.tlsKey;
+      "ssl/glusterfs.ca".source = cfg.tlsSettings.caCert;
```
OK, here I'm not quite sure what `source` exactly does.

Does it just point the symlink `/etc/ssl/glusterfs.key -> theGivenPath`? Or does it take a local path to a file, put it into the nix store, and create a symlink `/etc/ssl/glusterfs.key -> /nix/store/thefile`?

The latter would be bad, because that would make the private key world readable in `/nix/store`.

We definitely need the former, so that e.g. in the case of nixops we can obtain a symlink like `/etc/ssl/glusterfs.key -> /var/run/keys/...` where the target directory is the keys directory managed by nixops and not part of the store.

Can anybody shed light on this?

bachp replied:
On my machine it just creates the files end up in the nix store :(

```
lrwxrwxrwx 1 root root 28 Sep 17 17:33 glusterfs.ca -> /etc/static/ssl/glusterfs.ca
lrwxrwxrwx 1 root root 29 Sep 17 17:33 glusterfs.key -> /etc/static/ssl/glusterfs.key
lrwxrwxrwx 1 root root 29 Sep 17 17:33 glusterfs.pem -> /etc/static/ssl/glusterfs.pem
```

where:

```
lrwxrwxrwx 1 root root 50 Jan  1  1970 glusterfs.ca -> /nix/store/n9h1w46qjcbsrqpr38i7n6vdqb8cfj16-ca.pem
lrwxrwxrwx 1 root root 61 Jan  1  1970 glusterfs.key -> /nix/store/8skzrsdyx46kjzpz8m5kvj6y9x4qhqyy-cleopatra-key.pem
lrwxrwxrwx 1 root root 57 Jan  1  1970 glusterfs.pem -> /nix/store/54y38wlzipf142mxzki2qh3sz4hp4zxz-cleopatra.pem
```

I'm not sure how to do this properly?

joachifm replied:
Is there a particular reason for requiring the use of `environment.etc`? Compared to, say, using `preStart` or a dedicated unit for this purpose?

bachp replied:
The original entry I have in my `configuration.nix`:

```
  services.glusterfs = {
    enable = true;
    tlsSettings = {
      tlsKey = /home/pascal/ca/cleopatra-key.pem;
      tlsPem = /home/pascal/ca/cleopatra.pem;
      caCert = /home/pascal/ca/ca.pem; 
    };
  };
```

joachifm replied:
What if you enclose the paths in quotes, so they're treated as string literals and not paths (which will be added to the store on eval)?

joachifm replied:
(My brief reading of the etc builder code suggests it'll take whatever you give it, so long as it looks like a filepath).

nh2 commented on nixos/modules/services/network-filesystems/glusterfs.nix:
```diff
@@ -70,11 +128,12 @@ in
         LimitNOFILE=65536;
         ExecStart="${glusterfs}/sbin/glusterd -p /run/glusterd.pid --log-level=${cfg.logLevel} ${toString cfg.extraFlags}";
-        KillMode="process";
+        KillMode="control-group";
```
I would not make this change part of this commit / PR, but do that separately.

I have a branch [nh2-glusterfs-service-improvements](https://github.com/nh2/nixpkgs/tree/nh2-glusterfs-service-improvements) that includes [this commit](https://github.com/nh2/nixpkgs/commit/1cdb0b1d2954b73394a2a17a2b59bdb46360e633) to handle this.

My plan is that after your PR here has landed, we can merge the remaining improvements from my branch, and then do the gluster 3.12 upgrade (#29062).

bachp replied:
Good point I will remove this.

nh2 replied:
> My plan is that after your PR here has landed, we can merge the remaining improvements from my branch

I've PR'd the changes here now: https://github.com/NixOS/nixpkgs/pull/29868

nh2 commented:
I've made a couple comments / requests, the most important one is to figure out whether or not the current approach puts the private key into the nix store, if yes we have to find a different way.

nh2 commented:
Also note for myself: Once we've figured it out, I need to update my https://github.com/nh2/nixops-gluster-example/blob/8a859f0a702efd38bef9a202e1b031f1db6a44d3/modules/glusterfs-SSL-setup.nix#L87 which definitely has it in `/nix/store`, which is bad.

bachp committed:
gluster service: use str instead of path for private key

This pervents the user from accidently commiting the key to the nix store.
If providing a path instead of a string.

joachifm commented on nixos/modules/services/network-filesystems/glusterfs.nix:
```diff
@@ -59,9 +59,9 @@ in
+            tlsKeyPath = mkOption {
               default = null;
-              type = types.path;
+              type = types.str;
```
This does not protect against accidentally copying things into the store. It still accepts path literals, as in `/my/secret.key` which will be copied on eval.

bachp commented:
@nh2 Thanks for your feedback. All comments are addressed.

joachifm commented:
Changing the option type won't protect against accidental copying.

bachp commented:
@joachifm How you mean that? Like this I'm not able to pass a path to the option?

joachifm commented:
Two things: first, any value accepted by `types.path` is also accepted by `types.str` (the latter is a superset of the former). Secondly, Nix copies stuff into the store as a side-effect of evaluating them. This eval is different from module eval and may occur before module type checking, so you could still end up copying into the Nix store even if the module type-check eventually fails.

joachifm commented:
What you really want is a Nix level mechanism for preventing copies, I don't think you can easily achieve this via module options.

joachifm commented:
See https://github.com/NixOS/nixpkgs/blob/master/lib/types.nix#L167 for the definition of the `types.path`, you'll note it's just a string that begins with '/'.

bachp commented:
Just to clarify if:
1.  `tlsKeyPath = /path/key.pem;`  
  a. copies the key to nix store if `types.path` 
  b. gives an error of type missmatch if `types.str`;
2.  `tlsKeyPath = "/path/key.pem";`  creates a symlink from `/etc/static/ssl/glusterfs.key` -> `/path/key.pem` for both cases

You are saying in that case 1b fails but still has copied the key to the store?

Is case 2 always "safe" if I give it a string? And 1 is always "unsafe"?

joachifm commented:
By my analysis, 1 is always unsafe, and 2 is always safe. The copying that occurs in 1 is unrelated to the declared option type.

joachifm commented:
Well, I should say, safe in terms of what the Nix evaluator will do, the module could of course copy stuff in a different way (including it in `writeText` or what have you).

joachifm commented:
To clarify, the copying I'm worried about is what occurs when Nix evaluates the value `/foo/bar/baz` versus what happens when it evaluates `"/foo/bar/baz"`.

bachp commented:
@joachifm @nh2 So from what I hear I say we can drop https://github.com/NixOS/nixpkgs/pull/27340/commits/40e122ebb3460e039cc2185ce7f1d0ef1806fb63 again as it doesn't add any safety.

joachifm commented:
@bachp yes, if safety is the only motivation for the change it can be dropped IMO.

joachifm commented:
I *think* the best you can do right now is warn/fail if a keyfile path points into the Nix store, as that indicates accidental capture. The key will have leaked but at least the user knows. Other than that I imagine you need a Nix level mechanism to really fix this.

bachp commented:
I dropped the second commit.

nh2 commented:
@bachp I think we should keep the commit.

There's an important distinction here:

If you use nixops, then as you describe in https://github.com/NixOS/nixpkgs/pull/27340#issuecomment-330079307, method (1) copies the value into your local nix store. But afterwards, the evaluation will fail, and nixops will not get a chance to copy the world-readable key into the nix store of the remote machine(s).

That is much better, because that way the key leakage is confined to your own machine (the one you run nixops on), where you can easily undo it by deleting the built store path, and does not get deployed into the world.

And, as @joachifm says, it will ensure that the the user knows that they messed up (even though they still have to manually remove the key from their store).

bachp commented:
@joachifm @nh2 I re-added the commit. I think it is good to go now.

joachifm commented:
Thank you

nh2 commented:
Thanks!

nh2 commented:
Posting a comment here that I had before in my custom config file:

> Note: If the /var/lib/glusterd/secure-access thing breaks in the future and we get
> 
> `error:1408F10B:SSL routines:SSL3_GET_RECORD:wrong version number`
>
> in the server log, then that means that nixpkgs has switched the prefix `localstatedir=/var` to something different; the `secure-access` file must be under that prefix.

Just in case somebody needs to google that error message in the future.
